### PR TITLE
RD2: New RecurringDonation and RecurringDonationSettings fields

### PR DESCRIPTION
**Pull request by meena-sfdc**
# Critical Changes

# Changes
New fields have been added for the Recurring Donations 2

# Issues Closed

# New Metadata
- npe03__Recurring_Donation__c.InstallmentFrequency__c
- npe03__Recurring_Donation__c.PaymentMethod__c
- npe03__Recurring_Donation__c.RecurringType__c
- npe03__Recurring_Donation__c.Status__c
- npe03__Recurring_Donations_Settings__c.InstallmentOppStageName__c
- npe03__Recurring_Donations_Settings__c.IsRecurringDonations2Enabled__c

# Deleted Metadata

### CODE REVIEW

**Comment by meena-sfdc** [comment]
**lurch: attach W-035558

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001C9t7QAC" target="_blank">W-035558</a>

File: src/objects/npe03__Recurring_Donations_Settings__c.object [hunk 1]
```diff
@@ -1,5 +1,13 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <CustomObject xmlns="http://soap.sforce.com/2006/04/metadata">
+    <fields>
+        <fullName>IsRecurringDonations2Enabled__c</fullName>
```
Comment (force2b) [inline src/objects/npe03__Recurring_Donations_Settings__c.object]: @sam-knox Theses two new settings fields aren't in the Quip doc, but they will also need review & help text. Thanks.

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b - Thanks for flagging those. I'll take a look.

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @meena-sfdc - These settings aren't in our Quip doc, so I'll add my suggested changes here:

Change <label> on line 7 to: "Enhanced Recurring Donations".

Add this <inlineHelpText> and <description> "This option enables Enhanced Recurring Donations, and offers you the latest Recurring Donation features. But it also changes how you might have worked with Recurring Donations in the past. For complete details, see the NPSP product documentation."

@force2b 
For StageName_Installment__c, is it possible for this to be a picklist? If a user has to enter text themselves, do they enter the label or API name for the picklist value? What happens if they enter an invalid value (or no value at all)?

Comment (force2b) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @sam-knox Custom Settings do not support pick list field types, however we will eventually build a custom UI for this field in NPSP Settings that will allow the user to select a Stage from a list of all active Stages.

Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: Good idea, @force2b

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b - ok, thanks Mike.

@meena-sfdc For the StageName_Installment__c field, add this help text and description:

"The Opportunity Stage that NPSP applies to each new installment Opportunity when you create a Recurring Donation."

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @meena-sfdc - Use this for the help text and description for the Enhanced Recurring Donations field:

"This option enables Enhanced Recurring Donations, and offers you the latest Recurring Donation features. &lt;b&gt;WARNING&lt;/b&gt;: When you enable this option, NPSP changes your existing Recurring Donation records and the Recurring Donations object to match the data model for Enhanced Recurring Donations. &lt;b&gt;This is an irreversible process&lt;/b&gt;. For complete details, see the NPSP product documentation."

This is an update from what I had posted earlier.

Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: Thanks for the updates, @sam-knox

Comment (RadSFDeveloper) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b , since we are deciding on labels and descriptions, perhaps we can revisit API names?

- `IsRecurringDonations2Enabled__c` might be changed to `IsEnhancementEnabled__c` ?
- Remove `_` from `StageName_Installment__c`  and rename it to `InstallmentOppStageName__c`?

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @RadSFDeveloper - Yes, if we can align API names and labels, let's do it!

Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b , not sure if I can add bold <b> tags in the description of this field, I am getting errors when I deploy to scratch org.

Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b and @sam-knox , I made changes as per your suggestion for description and helptext and api names. However, the helptext for Enhanced Recurring Donations is exceeding the character limit. It does not say error, however, it is getting trimmed.

File: src/objects/npe03__Recurring_Donations_Settings__c.object [hunk 2]
```diff
@@ -1,5 +1,27 @@
...
+        <inlineHelpText>The Opportunity Stage that NPSP applies to each new installment Opportunity when you create a Recurring Donation.</inlineHelpText>
+        <label>Stage Name for Installment Opportunities</label>
+        <length>50</length>
+        <required>false</required>
+        <trackTrending>false</trackTrending>
+        <type>Text</type>
+        <unique>false</unique>
+    </fields>
+    <fields>
+        <fullName>IsEnhancementEnabled__c</fullName>
```
Comment (force2b) [inline src/objects/npe03__Recurring_Donations_Settings__c.object]: @sam-knox I'm not so sure about this field name to define when RD2 is enabled. The field name begs the question "What Enhancement". How about `IsEnhancedRDEnabled__c`?

Comment (RadSFDeveloper) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: The field is already in `npe03__Recurring_Donations_Settings__c` context so Label and Description give enough information, unless we are going to have another RD enhancement. However I have no objections to `IsEnhancedRDEnabled__c`.

File: src/objects/npe03__Recurring_Donation__c.object [hunk 3]
```diff
@@ -188,6 +188,105 @@
             </valueSetDefinition>
         </valueSet>
     </fields>
+    <fields>
+        <fullName>InstallmentFrequency__c</fullName>
+        <description>Defines the installment schedule of this Recurring Donation along with Installment Period. For example, if Installment Period is Monthly and Installment Frequency is 3, the schedule would be defined as &quot;every 3 months&quot;.</description>
+        <externalId>false</externalId>
+        <inlineHelpText>Defines the installment schedule of this Recurring Donation along with Installment Period. For example, if Installment Period is Monthly and Installment Frequency is 3, the schedule would be defined as &quot;every 3 months&quot;.</inlineHelpText>
+        <label>Installment Frequency</label>
+        <precision>18</precision>
```
Comment (force2b) [inline src/objects/npe03__Recurring_Donation__c.object]: @sam-knox What would you think about making this field a size of 3 so that cannot support a value beyond 999? As an Interval, it's meant to define something like "Every 3 Months" or "Every 90 Days". I can't see there ever being a case where we'd want a value over 100, but I'm hesitant to set a max size of 2.

File: src/objects/npe03__Recurring_Donations_Settings__c.object [hunk 4]
```diff
@@ -1,5 +1,27 @@
...
+        <defaultValue>false</defaultValue>
+        <description>This option enables Enhanced Recurring Donations, and offers you the latest Recurring Donation features. &lt;b&gt;WARNING&lt;/b&gt;: When you enable this option, NPSP changes your existing Recurring Donation records and the Recurring Donations object to match the data model for Enhanced Recurring Donations. &lt;b&gt;This is an irreversible process&lt;/b&gt;. For complete details, see the NPSP product documentation.</description>
```
Comment (force2b) [inline src/objects/npe03__Recurring_Donations_Settings__c.object]: @meena-sfdc HTML tags are not supported in HelpText or Description, so these can be removed.

Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: Thanks, @force2b

Comment (RadSFDeveloper) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b , do we plan to display field's Help Text on any custom page?

File: src/objects/npe03__Recurring_Donation__c.object (hunk 3 above)
Comment (sam-knox) [reply src/objects/npe03__Recurring_Donation__c.object]: @force2b - Yeah, that's a good call. 3 sounds like a good number. Then users can indicate "every 180 days" or whatever if they need to. We probably don't need to support "every 78,560 days" though!

File: src/objects/npe03__Recurring_Donations_Settings__c.object (hunk 1 above)
Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b - Is it possible to implement this using a custom label and keep the text? If not, I can try to shorten the text.

File: src/objects/npe03__Recurring_Donation__c.object (hunk 3 above)
Comment (force2b) [reply src/objects/npe03__Recurring_Donation__c.object]: Thanks @sam-knox 
@meena-sfdc Can you make this change as well?

File: src/objects/npe03__Recurring_Donations_Settings__c.object (hunk 4 above)
Comment (force2b) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: NPSP Settings page do sometimes display the InlineHelpText value, but we also use custom labels. In the case of this particular field though, there is no plan to render it directly in the NPSP Settings UI. There will be an enablement page, but it will not be a checkbox.

File: src/objects/npe03__Recurring_Donations_Settings__c.object (hunk 1 above)
Comment (force2b) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: I think this issue is settled now by updating the text in the UI and pulling it in, plus we've removed the html tags because they're not generally supported in InlineHelpText. The custom UI that will be built will not actually render the checkbox on the screen for enabling.

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: @force2b - Ok, so I don't need to try to shorten the text to under 255?

Comment (force2b) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: It should be good, thanks.
Also, I think that what's showing up in the "Conversation" tab in the PR is out of date. I'm going to resolve this conversation since it seems not match the actual file contents in the "Files Changed" tab.

File: src/objects/npe03__Recurring_Donation__c.object (hunk 3 above)
Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donation__c.object]: The code review changes are done and ready for review. The helptext for Enhanced recurring donations needs to be decreased in characters as it exceeds the limit.

Comment (sam-knox) [reply src/objects/npe03__Recurring_Donation__c.object]: @meena-sfdc - Please use this text for the Enhanced Recurring Donations Help Text and Description. It's under the character limit now:

"This option enables Enhanced Recurring Donations. WARNING: When you enable this option, NPSP changes your existing Recurring Donations records and the Recurring Donations object to match the data model for Enhanced Recurring Donations."

File: src/objects/npe03__Recurring_Donations_Settings__c.object [hunk 5]
```diff
@@ -1,5 +1,27 @@
...
+        <inlineHelpText>The Opportunity Stage that NPSP applies to each new installment Opportunity when you create a Recurring Donation.</inlineHelpText>
+        <label>Stage Name for Installment Opportunities</label>
+        <length>50</length>
+        <required>false</required>
+        <trackTrending>false</trackTrending>
+        <type>Text</type>
+        <unique>false</unique>
+    </fields>
+    <fields>
+        <fullName>IsEnhancementRDEnabled__c</fullName>
```
Comment (force2b) [inline src/objects/npe03__Recurring_Donations_Settings__c.object]: @meena-sfdc and @RadSFDeveloper I've been thinking about this field api name more, and after conferring with Mary and Joy, I'd like to go with field api name of `IsRecurringDonations2Enabled__c`. The label can remain to what Sam has approved. The two do not need to be in sync. From a system level this is "Recurring Donations 2.0" though we're communicating it as "Enhanced Recurring Donations". There may come a time when we eventually take RD another step further (RD3 for example) or even have another level of "enhanced". Using the "RecurringDonations2" in the api name allows us the option of changing the marketing label at any point while still ensuring a consistent use of the Api Name in the code (for example, the FeatureParameter for this is `IsEnabled_RecurringDonations2`).
CC @sam-knox @npsp-reedestockton

Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donations_Settings__c.object]: Thanks for the confirmation, @force2b . The changes have been made and ready for review.

File: src/objects/npe03__Recurring_Donation__c.object (hunk 3 above)
Comment (meena-sfdc) [reply src/objects/npe03__Recurring_Donation__c.object]: Thanks, @sam-knox . I have made the changes and it looks good now.

File: src/package.xml [hunk 6]
```diff
@@ -1190,7 +1190,13 @@
...
         <members>npe03__Recurring_Donation__c.Always_Use_Last_Day_Of_Month__c</members>
         <members>npe03__Recurring_Donation__c.Day_of_Month__c</members>
-        <members>npe03__Recurring_Donations_Settings__c.Recurring_Donation_Batch_Size__c</members>
+        <members>npe03__Recurring_Donation__c.InstallmentFrequency__c</members>        
+        <members>npe03__Recurring_Donation__c.PaymentMethod__c</members>
+        <members>npe03__Recurring_Donation__c.RecurringType__c</members>
+        <members>npe03__Recurring_Donation__c.Status__c</members>
+        <members>npe03__Recurring_Donations_Settings__c.IsRecurringDonations2Enabled__c</members>
+        <members>npe03__Recurring_Donations_Settings__c.Recurring_Donation_Batch_Size__c</members>        
+        <members>npe03__Recurring_Donations_Settings__c.InstallmentOppStageName__c</members>
```
Comment (force2b) [inline src/package.xml]: This line should move up one so the fields are in alphabetical order for the object.

Comment (meena-sfdc) [reply src/package.xml]: The changes are done and ready for review, thanks, @force2b !

**Comment by sam-knox** [comment]
@force2b @mpusto @meena-sfdc - Based on our conversation today about the enablement workflow, I'd like to update the Help Text and Description for IsRecurringDonations2Enabled__c. I think we should remove the "warning" in the current text because just checking the box doesn't rearrange any data or deploy anything.

New text:

"This option enables Enhanced Recurring Donations."

**Comment by force2b** [comment]
@sam-knox Very good point. Though, what would you think about having text such as "Warning: To avoid data inconsistencies, do not modify this field outside of NPSP Settings"? Basically I'm hoping to warn off any brave admins from manually checking that box in Custom Settings without going through our proper enablement process.

**Comment by sam-knox** [comment]
@force2b - How about this wording?

"This option enables Enhanced Recurring Donations. IMPORTANT: You must enable Enhanced Recurring Donations in NPSP Settings. Enabling in Custom Settings may cause data inconsistencies."

**Comment by force2b** [comment]
@sam-knox Very nice!

**Comment by meena-sfdc** [comment]
@sam-knox and @force2b , The changes are done!

**Comment by rponti-sforg** [comment]
QE testing for this PR is complete.  Tests have passed.  

For further testing details, see: https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000001C9t7QAC/view
https://salesforce.quip.com/JHhPAa9bLa7L

Merging PR and deleting branch.
